### ISSUE TITLE
Support generics, enums with non-unit variants, and opaque structs.

### ISSUE THREAD

**Issue by Sean1708** [Sean1708/rusty-cheddar#17]
`rustc` is absolutely fine with adding `#[repr(C)]` to generic structs, generic enums, and enums with non-unit variants so it's perfectly possible that these types might map nicely to C.

**Comment by Sean1708** [comment]
See [here](https://doc.rust-lang.org/nomicon/data.html) for info on representation.

**Comment by Sean1708** [comment]
The docs suggest [using an empty enum for opaque structs](http://doc.rust-lang.org/nightly/book/ffi.html#representing-opaque-structs), but these [can't be `#[repr(C)]`](https://play.rust-lang.org/?gist=19066e2b2a6af9e561ff&version=stable).

What about unit structs?

**Comment by Sean1708** [comment]
Even the enum would be easy enough to support, just check the variants are empty before checking for `#[repr(C)]` and special case that.

**Comment by Sean1708** [comment]
Thinking about it, you can't use empty enums for opaque structs because they need to be not opaque on the Rust side.

Instead I think private `#[repr(C)]` structs should be opaque structs.

**Comment by Sean1708** [comment]
The two current thoughts for opaque structs are

```rust
struct NonOpaque2 {
    b: f64,
}

#[repr(C)]
pub struct Opaque2(NonOpaque2);
```

À la [this playground](https://play.rust-lang.org/?gist=9907f7c7bafa5e57bb83&version=stable). So any public `#[repr(C)]` newtype would be an opaque struct that would have to be handled via pointers in C.


Alternatively have any public `#[repr(C)]` struct with _only_ private members be an opaque struct, this feels too magic to me though.

**Comment by Sean1708** [comment]
Closed by b6a7cc1ebe6a2d741eea64cfe4cecde60deeddc6
